contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,626 @@
+scores from all plugins according to the configured plugin weights.
 
-### Post-scoring/pre-reservation
+If a scoring plugin returns an error, the scheduler will treat it as a zero
+score.
 
-After all scoring plugins are invoked and the score of nodes are determined, the
-framework picks the best node with the highest score and then it calls
-post-scoring plugins. The Pod and the chosen Node are passed to these plugins.
-These plugins have one more chance to check any conditions about the assignment
-of the Pod to this Node and reject the node if needed.
+### Normalize scoring
 
-![image](images/scheduling-framework-threads.png)
+These plugins are used to modify scores before the scheduler computes a final
+ranking of Nodes. A plugin that registers for this extension point will be
+called with the [scoring](#scoring) results from the same plugin. This is called
+once per plugin per scheduling cycle.
+
+For example, suppose a plugin `BlinkingLightScorer` ranks Nodes based on how
+many blinking lights they have.
+
+```go
+func ScoreNode(_ *v1.Pod, n *v1.Node) (int, error) {
+   return getBlinkingLightCount(n)
+}
+```
+
+However, the maximum count of blinking lights may be small compared to
+`NodeScoreMax`. To fix this, `BlinkingLightScorer` should also register for this
+extension point.
+
+```go
+func NormalizeScores(scores map[string]int) {
+   highest := 0
+   for _, score := range scores {
+      highest = max(highest, score)
+   }
+   for node, score := range scores {
+      scores[node] = scores*NodeScoreMax/highest
+   }
+}
+```
+
+If any normalize-scoring plugin returns an error, the scheduling cycle is
+aborted.
+
+**Note:** Plugins wishing to perform "pre-reserve" work should use the
+normalize-scoring extension point.
 
 ### Reserve
 
-At this point scheduler updates its cache by "reserving" a Node (partially or
-fully) for the Pod. In scheduler v1 this stage is called "assume".
-At this point, only the scheduler cache is updated to
-reflect that the Node is (partially) reserved for the Pod. The scheduling
-framework calls plugins registered at this extension points so that they get a
-chance to perform cache updates or other accounting activities. These plugins
-do not return any value (except errors).
+This is an informational extension point. Plugins which maintain runtime state
+(aka "stateful plugins") should use this extension point to be notified by the
+scheduler when resources on a node are being reserved for a given Pod. This
+happens before the scheduler actually binds the pod to the Node, and it exists
+to prevent race conditions while the scheduler waits for the bind to succeed.
 
-The actual assignment of the Node to the Pod happens during the "Bind" phase.
-That is when the API server updates the Pod object with the Node information.
+Once a pod is in the reserved state, it will either trigger
+[Un-reserve](#un-reserve) plugins (on failure) or [Post-bind](#post-bind)
+plugins (on success).
```

Comment:
If we reserved the pod partially (e.g. some succeeded, some failed), I guess we should trigger un-reserve plugins too. What do you think?

For example, if volume scheduling plugin reserved some PVs for a pod, but another plugin failed to reserve other resources. If we don't un-reserve in this case and this pod is stuck in reserve phase because other resource it requests cannot be reserved, these PVs cannot be used by other pods anymore.

Reply:
I agree that, once a pod enters the reserve phase, we must either end up calling un-reserve or post-bind.

For the volume binder example, volume provisioning probably should happen in pre-bind and not in the reserve step. I can add some clarification to the extension points to make this more obvious.